File: client/src/models/devis.js
/**
 * Modèle de l'entité Devis.
 */
export const DEVIS_FIELDS = {
  id: 'number',
  reference: 'string',
  prospect_id: 'number',
  prospect_name: 'string|null',
  prospect_company: 'string|null',
  titre: 'string',
  description: 'string|null',
  montant: 'number',
  arr: 'number (récurrent annuel, 0 si aucun abonnement)',
  renewal_date: 'string|null (date de renouvellement)',
  items: 'string (JSON: {name, qty, price, period|null})',
  statut: 'string (brouillon | soumis | valide | refuse)',
  created_by: 'number',
  validated_by: 'number|null',
  validation_comment: 'string|null',
  created_at: 'string (date ISO)',
  archived_at: 'string|null (devis archivé)',
  updated_at: 'string (date ISO)',
};

/**
 * Fabrique un devis avec des valeurs par défaut.
 * @param {object} data - valeurs initiales (optionnel)
 * @returns {object} une structure devis
 */
export function createDevis(data = {}) {
  return {
    id: null,
    reference: '',
    prospect_id: null,
    prospect_name: null,
    prospect_company: null,
    titre: '',
    description: null,
    montant: 0,
    arr: 0,
    renewal_date: null,
    items: null,
    statut: 'brouillon',
    created_by: null,
    validated_by: null,
    validation_comment: null,
    created_at: null,
    updated_at: null,
    archived_at: null,
    ...data,
  };
}

export default createDevis;

File: mobile/lib/models/devis.dart
import '../utils/formatters.dart';

class DevisItem {
  final String name;
  final int qty;
  final double price;
  final String? period;

  const DevisItem({required this.name, this.qty = 1, this.price = 0, this.period});

  factory DevisItem.fromJson(Map<String, dynamic> json) {
    return DevisItem(
      name: json['name'] as String? ?? '',
      qty: toInt(json['qty']) ?? 1,
      price: toDouble(json['price']) ?? 0,
      period: json['period'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'name': name,
        'qty': qty,
        'price': price,
        if (period != null) 'period': period,
      };

  double get total => qty * price;
  double get annual => total * switch (period) {
        'mensuel' => 12,
        'trimestriel' => 4,
        'annuel' => 1,
        _ => 0,
      };
}

class Devis {
  final int id;
  final String reference;
  final int? prospectId;
  final String? prospectName;
  final String? prospectCompany;
  final String titre;
  final String? description;
  final double montant;
  final double arr;
  final String statut;
  final int? createdBy;
  final String? renewalDate;
  final String? createdByName;
  final String? validatedByName;
  final String? validationComment;
  final List<DevisItem> items;
  final String? createdAt;

  const Devis({
    required this.id,
    required this.reference,
    this.prospectId,
    this.prospectName,
    this.prospectCompany,
    required this.titre,
    this.description,
    this.montant = 0,
    this.arr = 0,
    this.statut = 'brouillon',
    this.createdBy,
    this.renewalDate,
    this.createdByName,
    this.validatedByName,
    this.validationComment,
    this.items = const [],
    this.createdAt,
  });

  factory Devis.fromJson(Map<String, dynamic> json) {
    return Devis(
      id: toInt(json['id']) ?? 0,
      reference: json['reference'] as String? ?? '',
      prospectId: toInt(json['prospect_id']),
      prospectName: json['prospect_name'] as String?,
      prospectCompany: json['prospect_company'] as String?,
      titre: json['titre'] as String? ?? '',
      description: json['description'] as String?,
      montant: toDouble(json['montant']) ?? 0,
      arr: toDouble(json['arr']) ?? 0,
      statut: json['statut'] as String? ?? 'brouillon',
      createdBy: toInt(json['created_by']),
      renewalDate: json['renewal_date'] as String?,
      createdByName: json['created_by_name'] as String?,
      validatedByName: json['validated_by_name'] as String?,
      validationComment: json['validation_comment'] as String?,
      items: (json['items'] as List?)
              ?.map((e) => DevisItem.fromJson(e as Map<String, dynamic>))
              .toList() ??
          const [],
      createdAt: json['created_at'] as String?,
    );
  }
}